edge/pkg/devicetwin/dtmanager/dmiserver.go

```diff
@@ -0,0 +1,427 @@
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+			err = GrpcCreateDeviceModel(&dm)
+			if err != nil {
+				klog.Errorf("add device model %s failed with err: %s", dm.Name, err)
+				return err
+			}
+
+			deviceModelList[dm.Name] = &dm
+		case model.DeleteOperation:
+			err := json.Unmarshal(message.Content.([]byte), &dm)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+			err = GrpcRemoveDeviceModel(&dm)
+			if err != nil {
+				klog.Errorf("delete device model %s failed with err: %s", dm.Name, err)
+				return err
+			}
+
+			delete(deviceModelList, dm.Name)
+		case model.UpdateOperation:
+			err := json.Unmarshal(message.Content.([]byte), &dm)
+			if err != nil {
+				return fmt.Errorf("invalid message content with err: %+v", err)
+			}
+			err = GrpcUpdateDeviceModel(&dm)
+			if err != nil {
+				klog.Errorf("update device model %s failed with err: %s", dm.Name, err)
+				return err
+			}
+
+			deviceModelList[dm.Name] = &dm
+		default:
+			klog.Warningf("unsupported operation %s", message.GetOperation())
+		}
+
+	default:
+		klog.Warningf("unsupported resource type %s", resources[3])
+	}
+
+	return nil
+}
+
+func GrpcCreateDevice(device *v1alpha2.Device) error {
+	protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+	if err != nil {
+		return err
+	}
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, err := dmiserver.GetDMIClientByProtocol(protocol)
+	if err != nil {
+		return err
+	}
+
+	cdr, err := dmiclient.CreateDeviceRequest(device)
+	if err != nil {
+		return fmt.Errorf("fail to create CreateDeviceRequest for device %s with err: %v", device.Name, err)
+	}
+	_, err = dc.Client.CreateDevice(dc.Ctx, cdr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcRemoveDevice(device *v1alpha2.Device) error {
+	protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+	if err != nil {
+		return err
+	}
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
+	if err != nil {
+		return err
+	}
+
+	defer conn.Close()
+	defer cancelFunc()
+
+	rdr, err := dmiclient.RemoveDeviceRequest(device.Name)
+	if err != nil {
+		return fmt.Errorf("fail to generate RemoveDeviceRequest for device %s with err: %v", device.Name, err)
+	}
+	_, err = dc.RemoveDevice(ctx, rdr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GrpcUpdateDevice(device *v1alpha2.Device) error {
+	protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+	if err != nil {
+		return err
+	}
+	sockPath := GetSockPath(protocol)
+	if sockPath == "" {
+		return fmt.Errorf("cannot get sockPath of protocol %s", protocol)
+	}
+
+	dc, ctx, conn, cancelFunc, err := dmiclient.GenerateDMIClient(sockPath)
```

Comment:
use GetDMIClientByProtocol to get client

Reply:
done